components/raftstore-v2/src/operation/ready/snapshot.rs

```diff
@@ -313,4 +339,64 @@ impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
         *snap_state = SnapState::Generated(snap);
         true
     }
+
+    pub fn on_applied_snapshot(&mut self) {
+        let mut entry = self.entry_storage_mut();
+        let term = entry.truncated_term();
+        let index = entry.truncated_index();
+        entry.set_applied_term(term);
+        entry.apply_state_mut().set_applied_index(index);
+        self.region_state_mut().set_tablet_index(index);
+    }
+
+    pub fn apply_snapshot(
+        &mut self,
+        snap: &Snapshot,
+        task: &mut WriteTask<EK, ER>,
+        snap_mgr: TabletSnapManager,
+        tablet_factory: Arc<dyn TabletFactory<EK>>,
+    ) -> Result<()> {
+        let region_id = self.region().get_id();
+        let peer_id = self.peer().get_id();
+        info!(
+            self.logger(),
+            "begin to apply snapshot";
+        );
+
+        let mut snap_data = RaftSnapshotData::default();
+        snap_data.merge_from_bytes(snap.get_data())?;
+        let region = snap_data.take_region();
+        if region.get_id() != region_id {
+            return Err(box_err!(
+                "mismatch region id {}!={}",
+                region_id,
+                region.get_id()
+            ));
+        }
+
+        let last_index = snap.get_metadata().get_index();
+        let last_term = snap.get_metadata().get_term();
+        self.region_state_mut().set_state(PeerState::Normal);
+        self.region_state_mut().set_region(region);
+        self.entry_storage_mut()
+            .raft_state_mut()
+            .set_last_index(last_index);
+        self.entry_storage_mut().set_truncated_index(last_index);
+        self.entry_storage_mut().set_truncated_term(last_term);
+        self.entry_storage_mut().set_last_term(last_term);
+
+        let key = TabletSnapKey::new(region_id, peer_id, last_term, last_index);
+        let mut path = snap_mgr.final_recv_path(&key);
+        let logger = self.logger().clone();
+        // The snapshot require no additional processing such as ingest them to DB, but
+        // it should load it into the factory after it persisted.
+        let hook = move || {
+            if let Err(e) = tablet_factory.load_tablet(path.as_path(), region_id, last_index) {
+                error!(logger, "failed to load tablet";"path" => path.display(),"err" => ?e);
```

Comment:
It should panic instead of just logging. Fail to load tablet means data is corrupted.

Reply:
fixed